pkg/rbac/userData.go

```diff
@@ -0,0 +1,147 @@
+// Copyright Contributors to the Open Cluster Management project
+package rbac
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/stolostron/search-v2-api/pkg/config"
+	authzv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+// Contains data about the resources the user is allowed to access.
+type userData struct {
+	// clusters     []string              // Managed clusters where the user has view access.
+	// csResources  []resource            // Cluster-scoped resources on hub the user has list access.
+	nsResources map[string][]resource // Namespaced resources on hub the user has list access.
+
+	// Internal fields to manage the cache.
+	// clustersErr       error      // Error while updating clusters data.
+	// clustersLock      sync.Mutex // Locks when clusters data is being updated.
+	// clustersUpdatedAt time.Time  // Time clusters was last updated.
+	err     error      // Error while getting user data from cache
+	csrErr  error      // Error while updating cluster-scoped resources data.
+	csrLock sync.Mutex // Locks when cluster-scoped resources data is being updated.
+	// csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
+	nsrErr       error      // Error while updating namespaced resources data.
+	nsrLock      sync.Mutex // Locks when namespaced resources data is being updated.
+	nsrUpdatedAt time.Time  // Time namespaced resources was last updated.
+
+	// impersonate *kubernetes.Interface // client with impersonation config
+}
+
+func (cache *Cache) GetUserData(ctx context.Context, clientToken string) (*userData, error) {
+	var user userData
+	uid := cache.tokenReviews[clientToken].tokenReview.Status.User.UID //get uid from tokenreview
+	cache.usersLock.Lock()
+	defer cache.usersLock.Unlock()
```

Comment:
Do we need the lock for looking up the cache , or we could live with locking only when we update the cache. ?

Reply:
This lock here is to lock the cache while we are initializing a new user.